telegram/ext/_picklepersistence.py

```diff
@@ -26,14 +28,99 @@
     Tuple,
     overload,
     cast,
+    Type,
+    Set,
+    Callable,
 )
 
+from telegram import Bot, TelegramObject
 from telegram._utils.types import FilePathInput
+from telegram._utils.warnings import warn
 from telegram.ext import BasePersistence, PersistenceInput
 from telegram.ext._contexttypes import ContextTypes
 from telegram.ext._utils.types import UD, CD, BD, ConversationDict, CDCData
 
 
+_REPLACED_KNOWN_BOT = "a known bot replaced by PTB's PicklePersistence"
+_REPLACED_UNKNOWN_BOT = "an unknown bot replaced by PTB's PicklePersistence"
+
+
+def _all_subclasses(cls: Type[TelegramObject]) -> Set[Type[TelegramObject]]:
+    """Gets all subclasses of the specified object, recursively. from
+    https://stackoverflow.com/a/3862957/9706202
+    """
+    subclasses = cls.__subclasses__()
+    return set(subclasses).union([s for c in subclasses for s in _all_subclasses(c)])
+
+
+def _reconstruct_to(cls: Type[TelegramObject], kwargs: dict) -> TelegramObject:
+    """
+    This method is used for unpickling. The data, which is in the form a dictionary, is
+    converted back into a class. Functions the same as :meth:`TelegramObject.__setstate__`
+    """
+    obj = cls(**kwargs)
+    if '_bot' in kwargs:
+        obj.set_bot(kwargs['_bot'])
+    return obj
+
+
+def _custom_reduction(cls: TelegramObject) -> Tuple[Callable, Tuple[Type[TelegramObject], dict]]:
+    """
+    This method is used for pickling. The bot attribute is preserved so _BotPickler().persistent_id
+    works as intended.
+    """
+    # Adapted from self.to_dict(), unlike to_dict(), this is not recursive.
+    data = {}
+    attrs = {attr for cls in cls.__class__.__mro__[:-1] for attr in cls.__slots__}
+    for key in attrs:
+        value = getattr(cls, key, None)
+        if value is not None:
+            data[key] = value
+
+    return _reconstruct_to, (cls.__class__, data)
+
+
+class _BotPickler(pickle.Pickler):
+    __slots__ = ('_bot',)
+
+    # Here we define a private dispatch_table, because we want to preserve the bot attribute of
+    # objects so persistent_id works as intended. Otherwise, the bot attribute is deleted in
+    # __getstate__, which used for regular pickling (via pickle.dumps(...))
+    dispatch_table = copyreg.dispatch_table.copy()  # type: ignore[attr-defined]
+    for obj in _all_subclasses(TelegramObject):
+        dispatch_table[obj] = _custom_reduction
```

Comment:
is there a specific reason why this is not in `__init__`? And We only need this in py3.7, right? If `reducer_override` is respected in py3.8+, I would suggest to place this in an `if py_version < (3,8)` clause and implement the `reducer_override` method in addition.
My gut feeling is that the `reducer_override` approach may be a bit more stable than the "get all subclasses" logic

Reply:
> why not in `__init__`?

Well we don't need to generate the dispatch_table again and again on `__init__`, doing it once is fine imo. 

And yes, I guess it makes sense to do the `reduce_override` thing for future proofing

Reply:
> > why not in `__init__`?
> 
> Well we don't need to generate the dispatch_table again and again on `__init__`, doing it once is fine imo.

What happens if I add a new subclass at runtime?

> And yes, I guess it makes sense to do the `reduce_override` thing for future proofing